prjxray/fasm_assembler.py

```diff
@@ -0,0 +1,148 @@
+import fasm
+from prjxray import bitstream
+from prjxray import grid
+
+
+class FasmLookupError(Exception):
+    pass
+
+
+class FasmInconsistentBits(Exception):
+    pass
+
+
+def init_frame_at_address(frames, addr):
+    '''Set given frame to 0 if not initialized '''
+    if not addr in frames:
+        frames[addr] = [0 for _i in range(bitstream.FRAME_WORD_COUNT)]
+
+
+class FasmAssembler(object):
+    def __init__(self, db):
+        self.db = db
+        self.grid = db.grid()
+
+        self.frames = {}
+        self.frames_line = {}
+
+    def get_frames(self, sparse=False):
+        if not sparse:
+            frames = self.frames_init()
+        else:
+            frames = {}
+
+        for (frame_addr, word_addr, bit_index), is_set in self.frames.items():
+            init_frame_at_address(frames, frame_addr)
+
+            if is_set:
+                frames[frame_addr][word_addr] |= 1 << bit_index
+
+        return frames
+
+    def frames_init(self):
+        '''Set all frames to 0'''
+        frames = {}
+
+        for bits_info in self.grid.iter_all_frames():
+            for coli in range(bits_info.bits.frames):
+                init_frame_at_address(
+                    frames, bits_info.bits.base_address + coli)
+
+        return frames
+
+    def frame_set(self, frame_addr, word_addr, bit_index, line):
+        '''Set given bit in given frame address and word'''
+        assert bit_index is not None
+
+        key = (frame_addr, word_addr, bit_index)
+        if key in self.frames:
+            if self.frames[key] != 1:
+                raise FasmInconsistentBits(
+                    'FASM line "{}" wanted to set bit {} but was cleared by FASM line "{}"'
+                    .format(
+                        line,
+                        key,
+                        self.frames_line[key],
+                    ))
+            return
+
+        self.frames[key] = 1
+        self.frames_line[key] = line
+
+    def frame_clear(self, frame_addr, word_addr, bit_index, line):
+        '''Set given bit in given frame address and word'''
+        assert bit_index is not None
+
+        key = (frame_addr, word_addr, bit_index)
+        if key in self.frames:
+            if self.frames[key] != 0:
+                raise FasmInconsistentBits(
+                    'FASM line "{}" wanted to clear bit {} but was set by FASM line "{}"'
+                    .format(
+                        line,
+                        key,
+                        self.frames_line[key],
+                    ))
+            return
+
+        self.frames[key] = 0
+        self.frames_line[key] = line
+
+    def enable_feature(self, tile, feature, address, line):
+        gridinfo = self.grid.gridinfo_at_tilename(tile)
+
+        # TODO: How to determine if the feature targets BLOCK_RAM segment type?
```

Comment:
not sure what the specific issue is here, but feel free to ping me if you need additional info on BRAM

Reply:
The original fasm2frames.py didn't seem to handle this either.  In the current database format, how do we know if a FASM line is targeting the CLB_IO_CLK section or the BLOCK_RAM section?  Maybe I missed the database file or something.  We could special case BLOCK_RAM FASM features as simply
BRAM_X<x>_Y<y>.BLOCK_RAM
etc.

Reply:
Ah yeah, this isn't fully hammered out yet, but I'm assuming something like that symbols don't overlap between the memory regions. ie something under the hood tags it as coming from a certain memory region, and fasm knows which region to map that to without the user specifying explicitly

Reply:
Hence why I left that as a TODO.  We'll need to add FASM support for BRAM initialization eventually, but I think we can leave that out of scope for this PR.